docs/proposals-accepted/202201-parallel-compaction.md

```diff
@@ -0,0 +1,123 @@
+title: Compaction of a single tenant in parallel
+status: accepted
+owner: roystchiang
+menu: proposals-accepted
+---
+
+* **Owners:**
+  * @roystchiang
+
+* **Related Tickets:**
+  * [Improvements on Thanos Compactor](https://github.com/thanos-io/thanos/issues/4233)
+
+## Summary
+
+This design document proposes the ability to produce parallelizable compaction plan and the ability to execute them in parallel.
+
+## Motivation
+
+The main motivation for considering parallel compaction in Thanos are the following use cases:
+
+* **A single tenant wants to push more timeseries:** One of the bottlenecks for running a Thanos cluster is how many timeseries can be compacted within the minimal compaction range. Using a compaction range of [2h, 8h, 2d, 25d]  as example, if the compaction for the 2h block is longer than 2 hours, compaction will start to fall behind.
+* **Faster recovery from failed compactor nodes:** In the event of compactor nodes experiencing failures, compaction has to be done for blocks that were uploaded during the outage. By compacting in non-overlapping plans in parallel, we can recover faster.
+
+## Goals
+
+* Ability to create parallelizable compaction plans
+* Ability to execute multiple compaction plans in parallel
+
+## Proposal
+
+Currently a single tenant belongs to a single compactor, and compactions for a single tenant happens serially. In reality, this compaction can be done in parallel, if the blocks to be compacted do not overlap. We can increase the throughput of compaction, with the ability to produce plans with groups of blocks, and compact them in parallel.
+
+## Proposed Approach
+
+The compactor can be split into two components - a planner and a worker.  The planner simply reads a list of blocks in the persistent storage, and generates compaction plans. The worker accepts plans from a planner, and downloads the blocks, compacts them, and uploads the new block.
+
+We’ll tackle the problem in two stages. First, enabling parallelism within a single compactor, then we’ll move the parallelism to a fleet of compaction workers. By introducing a grouper that’s capable of producing multiple plans, we can spin up a thread for each plan, and start compacting. Since each tenant still belongs to a single compactor, there is no coordination involved across compactors. This alone will allow compactor to scale vertically, and users can already benefit from it. However, to achieve horizontal scaling, we’ll need the ability to split work for a single tenant to multiple compactors.
+
+We’ll introduce the ability for compactor to be running in planner, worker, and planner + worker mode. By running the compactor in a planner mode, the compactor will scan the object store for a list of blocks, and produce plans. The worker nodes will receive jobs from the planner nodes, and execute the compaction. Existing users do not have to spin up a new component, as they can still enjoy parallelism by vertically scaling their node.
```

Comment:
What' s the `planner + worker` mode? Is it the same as the current thanos compactor?

Reply:
With the ability to provide parallelizable plan for a single block stream, the current Thanos compactor has to scale up vertically..

The proposed `planner + worker` mode moves the parallelism to across multiple compactors, so now we can scale horizontally instead.

I've updated the proposal with better description. Thanks.